File: frontend/src/types/graph.ts
export type ResourceKind =
  | 'Pod'
  | 'Deployment'
  | 'StatefulSet'
  | 'DaemonSet'
  | 'Job'
  | 'CronJob'
  | 'Service'
  | 'Ingress'
  | 'ConfigMap'
  | 'Secret'
  | 'PersistentVolumeClaim'
  | 'PersistentVolume'
export type ViewType = 'graph' | 'bubbles' | 'table'

export interface ServicePort {
  name?: string
  port: number
  target_port?: string
  protocol: string
}

export interface IngressRule {
  host?: string
  path: string
  service_name: string
  service_port?: string
}

export interface GraphNode {
  id: string
  kind: ResourceKind
  name: string
  namespace: string
  labels: Record<string, string>
  status?: string
  replicas?: number
  selector?: Record<string, string>
  ports?: ServicePort[]
  ingress_rules?: IngressRule[]
}

export interface GraphEdge {
  id: string
  source: string
  target: string
  relation: 'selects' | 'routes-to' | 'uses-config' | 'uses-secret' | 'uses-tls' | 'owns' | 'mounts' | 'bound-to'
}

export interface GraphResponse {
  namespace: string
  nodes: GraphNode[]
  edges: GraphEdge[]
  generated_at: string
}

export interface NamespacesResponse {
  namespaces: string[]
}

File: backend/app/models/graph.py
from datetime import datetime
from enum import Enum
from typing import Optional
from pydantic import BaseModel


class ResourceKind(str, Enum):
    Pod = "Pod"
    Deployment = "Deployment"
    StatefulSet = "StatefulSet"
    DaemonSet = "DaemonSet"
    Service = "Service"
    Ingress = "Ingress"
    ConfigMap = "ConfigMap"
    Secret = "Secret"


class ServicePort(BaseModel):
    name: Optional[str] = None
    port: int
    target_port: Optional[str] = None
    protocol: str = "TCP"


class IngressRule(BaseModel):
    host: Optional[str] = None
    path: str
    service_name: str
    service_port: Optional[str] = None


class GraphNode(BaseModel):
    id: str
    kind: ResourceKind
    name: str
    namespace: str
    labels: dict[str, str] = {}
    status: Optional[str] = None
    replicas: Optional[int] = None
    selector: Optional[dict[str, str]] = None
    ports: Optional[list[ServicePort]] = None
    ingress_rules: Optional[list[IngressRule]] = None


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    relation: str


class GraphResponse(BaseModel):
    namespace: str
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    generated_at: datetime
